crates/h264/src/sps.rs

```diff
@@ -53,179 +647,412 @@ impl Sps {
+                    num_units_in_tick,
+                    time_scale,
+                })
             }
         }
 
         Ok(Sps {
+            forbidden_zero_bit,
+            nal_ref_idc,
+            nal_unit_type: NALUnitType(nal_unit_type),
             profile_idc,
+            constraint_set0_flag,
+            constraint_set1_flag,
+            constraint_set2_flag,
+            constraint_set3_flag,
+            constraint_set4_flag,
+            constraint_set5_flag,
             level_idc,
+            seq_parameter_set_id,
             ext: sps_ext,
-            width,
-            height,
-            frame_rate,
+            log2_max_frame_num_minus4,
+            pic_order_cnt_type,
+            log2_max_pic_order_cnt_lsb_minus4,
+            pic_order_cnt_type1,
+            max_num_ref_frames,
+            gaps_in_frame_num_value_allowed_flag,
+            pic_width_in_mbs_minus1,
+            pic_height_in_map_units_minus1,
+            mb_adaptive_frame_field_flag,
+            direct_8x8_inference_flag,
+            frame_crop_info,
+            sample_aspect_ratio,
+            overscan_appropriate_flag,
             color_config,
+            chroma_sample_loc,
+            timing_info,
+        })
+    }
+
+    /// The height as a u64. This is computed from other fields, and isn't directly set.
+    ///
+    /// `height = ((2 - frame_mbs_only_flag as u64) * (pic_height_in_map_units_minus1 + 1) * 16) -
+    /// frame_crop_bottom_offset * 2 - frame_crop_top_offset * 2`
+    ///
+    /// We don't directly store `frame_mbs_only_flag` since we can tell if it's set:
+    /// If `mb_adaptive_frame_field_flag` is None, then `frame_mbs_only_flag` is set (1).
+    /// Otherwise `mb_adaptive_frame_field_flag` unset (0).
+    pub fn height(&self) -> u64 {
+        let base_height =
+            (2 - self.mb_adaptive_frame_field_flag.is_none() as u64) * (self.pic_height_in_map_units_minus1 + 1) * 16;
+
+        self.frame_crop_info.as_ref().map_or(base_height, |crop| {
+            base_height - (crop.frame_crop_top_offset + crop.frame_crop_bottom_offset) * 2
+        })
+    }
+
+    /// The width as a u64. This is computed from other fields, and isn't directly set.
+    ///
+    /// `width = ((pic_width_in_mbs_minus1 + 1) * 16) - frame_crop_right_offset * 2 - frame_crop_left_offset * 2`
+    pub fn width(&self) -> u64 {
+        let base_width = (self.pic_width_in_mbs_minus1 + 1) * 16;
+
+        self.frame_crop_info.as_ref().map_or(base_width, |crop| {
+            base_width - (crop.frame_crop_left_offset + crop.frame_crop_right_offset) * 2
+        })
+    }
+
+    /// Returns the frame rate as a f64.
+    ///
+    /// If `timing_info_present_flag` is set, then the `frame_rate` will be computed, and
+    /// if `num_units_in_tick` is nonzero, then the framerate will be:
+    /// `frame_rate = time_scale as f64 / (2.0 * num_units_in_tick as f64)`
+    pub fn frame_rate(&self) -> f64 {
```

Comment:
Should this be a `Option<f64>`, i am not sure. We are using 0.0 as a sentinel value (for not present) in rust which is an antipattern.